File: VS/lab3/Equipment.cpp
#include "Equipment.hpp"
#include "WarehouseEntry.hpp"

Equipment::Equipment(int id, std::string name, int count)
    : WarehouseEntry<int>(id, name), m_count(count) {}

int Equipment::GetCount() const {
    return m_count;
}

bool Equipment::IsEmpty() const {
    return m_count <= 0;
}

void Equipment::Increase(int count) {
    m_count += count;
}

int Equipment::Decrease(int count) {
    int taken = std::min(m_count, count);
    m_count -= taken;
    return taken;
}

File: lab5/Java/src/warehouse/equipments/Equipment.java
package warehouse.equipments;

import request.SupplyRequestDetail;
import warehouse.WarehouseEntry;

public abstract class Equipment extends WarehouseEntry<Integer> {

    int count;

    public static final Integer MIN_COUNT = 1;

    public Equipment(int id, String name, int count) { 
        super(id, name);
        if(count < MIN_COUNT){
            throw new IllegalArgumentException("Недопустимое количество");
        }         
        this.count = count;
    }

    public int getCount(){
        return count;
    }
    @Override
    public void increase(Number value){
        count+=value.intValue();
    }

    @Override
    public Integer decrease(Number value){
        int taken = Math.min(count, value.intValue());
        count-=taken;
        return taken;
    }
    
    @Override
    public Boolean isEmpty(){
        return count < MIN_COUNT;
    }

    @Override
    public abstract Boolean isMatches(SupplyRequestDetail detail);
}